kernel/hConstr.ml

```diff
@@ -0,0 +1,370 @@
+  | Prod (na,t,c) -> combinesmall 4 (combine3 (hash_annot na) t.hash c.hash)
+  | Lambda (na,t,c) -> combinesmall 5 (combine3 (hash_annot na) t.hash c.hash)
+  | LetIn (na,b,t,c) -> combinesmall 6 (combine4 (hash_annot na) b.hash t.hash c.hash)
+  | App (h,args) -> combinesmall 7 (Array.fold_left (fun hash c -> combine hash c.hash) h.hash args)
+  | Evar _ -> assert false
+  | Const (c,u) -> combinesmall 9 (combine (Constant.SyntacticOrd.hash c) (UVars.Instance.hash u))
+  | Ind (ind,u) -> combinesmall 10 (combine (Ind.SyntacticOrd.hash ind) (UVars.Instance.hash u))
+  | Construct (c,u) -> combinesmall 11 (combine (Construct.SyntacticOrd.hash c) (UVars.Instance.hash u))
+  | Case (_,u,pms,(p,_),_,c,bl) ->
+    let hash_ctx (bnd,c) =
+      Array.fold_left (fun hash na -> combine (hash_annot na) hash) c.hash bnd
+    in
+    let hpms = hash_array hash pms in
+    let hbl = hash_array hash_ctx bl in
+    let h = combine5 (UVars.Instance.hash u)
+        c.hash hpms (hash_ctx p) hbl
+    in
+    combinesmall 12 h
+  | Fix (_,(lna,tl,bl)) ->
+    combinesmall 13 (combine3 (hash_array hash_annot lna) (hash_array hash tl) (hash_array hash bl))
+  | CoFix (_,(lna,tl,bl)) ->
+    combinesmall 14 (combine3 (hash_array hash_annot lna) (hash_array hash tl) (hash_array hash bl))
+  | Meta _ -> assert false
+  | Rel n -> combinesmall 16 n
+  | Proj (p,_,c) -> combinesmall 17 (combine (Projection.SyntacticOrd.hash p) c.hash)
+  | Int i -> combinesmall 18 (Uint63.hash i)
+  | Float f -> combinesmall 19 (Float64.hash f)
+  | String s -> combinesmall 20 (Pstring.hash s)
+  | Array (u,t,def,ty) -> combinesmall 21 (combine4 (UVars.Instance.hash u) (hash_array hash t) def.hash ty.hash)
+
+let kind_to_constr = function
+  | Rel n -> mkRel n
+  | Var i -> mkVar i
+  | Meta _ | Evar _ -> assert false
+  | Sort s -> mkSort s
+  | Cast (c,k,t) -> mkCast (c.self,k,t.self)
+  | Prod (na,t,c) -> mkProd (na,t.self,c.self)
+  | Lambda (na,t,c) -> mkLambda (na,t.self,c.self)
+  | LetIn (na,b,t,c) -> mkLetIn (na,b.self,t.self,c.self)
+  | App (h,args) -> mkApp (h.self, Array.map self args)
+  | Const cu -> mkConstU cu
+  | Ind iu -> mkIndU iu
+  | Construct cu -> mkConstructU cu
+  | Case (ci,u,pms,(p,r),iv,c,bl) ->
+    let to_ctx (bnd,c) = bnd, c.self in
+    let iv = match iv with
+      | NoInvert -> NoInvert
+      | CaseInvert x -> CaseInvert {indices=Array.map self x.indices}
+    in
+    mkCase (ci,u,Array.map self pms,(to_ctx p,r),iv,c.self,Array.map to_ctx bl)
+  | Fix (ln,(lna,tl,bl)) ->
+    mkFix (ln,(lna,Array.map self tl,Array.map self bl))
+  | CoFix (ln,(lna,tl,bl)) ->
+    mkCoFix (ln,(lna,Array.map self tl,Array.map self bl))
+  | Proj (p,r,c) -> mkProj (p,r,c.self)
+  | Int i -> mkInt i
+  | Float f -> mkFloat f
+  | String s -> mkString s
+  | Array (u,t,def,ty) -> mkArray (u,Array.map self t,def.self,ty.self)
+
+let hcons_inplace f a = Array.iteri (fun i x -> Array.unsafe_set a i (f x)) a
```

Comment:
I know the kernel does that already, but I do find the in-place modification of arrays to be a potential problem. We may be silently breaking invariants, especially if we start using OCaml 5.0 parallelism. Is this really critical for performance?